### PR TITLE
Fixed #25563 -- Made sure deferred models use their proxied model app cache.

### PR DESCRIPTION
**Pull request by charettes**
Thanks to Andriy Sokolovskiy for the report.

### CODE REVIEW

**Comment by coldmind** [comment]
:+1:

File: tests/defer_regress/tests.py [hunk 1]
```diff
@@ -263,6 +265,24 @@ def test_deferred_class_factory_no_attrs(self):
         deferred_cls = deferred_class_factory(Item, ())
         self.assertFalse(deferred_cls._deferred)
 
+    def test_deferred_class_factory_apps_reuse(self):
+        """
+        #25563 - The model apps should be used for caching and
+        retrieval of the proxy class created.
```
Comment (timgraham) [inline tests/defer_regress/tests.py]: "of the proxy class created" -> "of the created proxy class"?

File: tests/defer_regress/tests.py [hunk 2]
```diff
@@ -263,6 +265,24 @@ def test_deferred_class_factory_no_attrs(self):
         deferred_cls = deferred_class_factory(Item, ())
         self.assertFalse(deferred_cls._deferred)
 
+    def test_deferred_class_factory_apps_reuse(self):
+        """
+        #25563 - The model apps should be used for caching and
```
Comment (timgraham) [inline tests/defer_regress/tests.py]: Not sure what "model apps" means? model._meta.apps?

Comment (charettes) [inline tests/defer_regress/tests.py]: Yes, what about "the apps the model is registered to"?

File: tests/defer_regress/tests.py (hunk 1 above)
Comment (charettes) [inline tests/defer_regress/tests.py]: Makes sense.

File: tests/defer_regress/tests.py (hunk 2 above)
Comment (timgraham) [inline tests/defer_regress/tests.py]: "model._meta.apps should be used..." seems clear to me.

**Comment by charettes** [comment]
@timgraham does the added release note and adjusted commit message make sense to you?

**Comment by timgraham** [comment]
The release note is a bit "low level" (i.e. it describes the fix but it's not really clear what issue it might solve) but I guess I don't mind in this case. At least the ticket link gives more context if a reader is interested.

**Comment by charettes** [comment]
I guess I could add a mention of the actual symptom. What about adding the following text:

> Before this change the deferred models used in data migrations would be leaked to test and application code.

**Comment by timgraham** [comment]
Sure, I'd just change the beginning to "This prevents any models generated in data migrations using QuerySet.defer() from leaking..."

**Comment by charettes** [comment]
Alright, just pushed the suggested changes.

**Pull request merged**